WEB-INTEGRATION.md

```diff
@@ -426,36 +426,226 @@ Event dispatches can be one of the following:
+      }
+    }
+  }
+}
+```
+
+An example _synchronous_ event is `AbortSignal`'s `abort` event. A naive approximation
+in JavaScript would look like the following:
+
+```javascript
+class AbortController {
+  constructor() {
+    this.signal = new AbortSignal();
+  }
+
+  abort() {
+    this.signal.aborted = true;
+    this.signal.dispatchEvent(new Event("abort"));
+  }
+}
+```
+
+When calling `abortController.abort()`, there is a current async context active in the agent. All operations that lead to the `abort` event being dispatched are synchronous and do not manually change the current async context: the active async context will remain the same through the whole `.abort()` process,
+including in the event listener callbacks:
+
+```javascript
+const abortController = new AbortController();
+const asyncVar = new AsyncContext.Variable();
+abortController.signal.addEventListener("abort", () => {
+  console.log(asyncVar.get()); // "foo"
+});
+asyncVar.run("foo", () => {
+  abortController.abort();
+});
+```
+
+Let's consider now a more complex case: the asynchronous `"load"` event of `XMLHttpRequest`. Let's try
+to implement `XMLHttpRequest` in JavaScript, on top of fetch:
+
+```javascript
+class XMLHttpRequest extends EventTarget {
+  #method;
+  #url;
+  open(method, url) {
+    this.#method = method;
+    this.#url = url;
+  }
+  send() {
+    (async () => {
+      try {
+        const response = await fetch(this.#url, { method: this.#method });
+        const reader = response.body.getReader();
+        let done;
+        while (!done) {
+          const { done: d, value } = await reader.read();
+          done = d;
+          this.dispatchEvent(new ProgressEvent("progress", { /* ... */ }));
+        }
+        this.dispatchEvent(new Event("load"));
+      } catch (e) {
+        this.dispatchEvent(new Event("error"));
+      }
+    })();
+  }
+}
+```
+
+And lets trace how the context propagates from `.send()` in the following case:
+```javascript
+const asyncVar = new AsyncContext.Variable();
+const xhr = new XMLHttpRequest();
+xhr.open("GET", "https://example.com");
+xhr.addEventListener("load", () => {
+  console.log(asyncVar.get()); // "foo"
+});
+asyncVar.run("foo", () => {
+  xhr.send();
+});
+```
+- when `.send()` is called, the value of `asyncVar` is `"foo"`.
+- it is synchronously propagated up to the `fetch()` call in `.send()`
+- the `await` snapshots the context before pausing, and restores it (to `asyncVar: "foo"`) when the `fetch` completes
+- the `await`s in the reader loop propagate the context as well
+- when `this.dispatchEvent(new Event("load"))`, is called, the current active async context is thus
+  the same one as when `.send()` was called
+- the `"load"` callback thus runs with `asyncVar` set to `"foo"`.
 
-### Fallback context
+Note that this example uses `await`, but due to the proposed semantics for `.then` and `setTimeout`
+(and similar APIs), the same would hapepn when using other asynchronicity primitives.
```

Comment:
Maybe add something like "If something isn't expressible using existing asynchronicity primitives, we assume that any internal primitives would take callbacks that would propagate the context as described above."

Things like "queue a task" are similar but not identical to `setTimeout(cb, 0)`, and if queuing a task in response to e.g. a websocket request, that couldn't exactly be expressed in terms of anything already in the language.

Reply:
To avoid "this is based on something that is not actually possible", I'll mention that you can still _emulate_ those API (e.g. you can mock the WebSocket API so that it looks like you are dealing with a real web socket, but you are actually not).